pixy/groupreg.go

```diff
@@ -0,0 +1,222 @@
+		groupMemberZNode:    groupMemberZNode,
+		topicsCh:            make(chan []string),
+		membershipChangesCh: make(chan map[string][]string),
+		stoppingCh:          make(chan none),
+	}
+	spawn(&cgr.wg, cgr.watcher)
+	spawn(&cgr.wg, cgr.register)
+	return cgr
+}
+
+func (cgr *consumerGroupRegistry) topics() chan<- []string {
+	return cgr.topicsCh
+}
+
+func (cgr *consumerGroupRegistry) membershipChanges() <-chan map[string][]string {
+	return cgr.membershipChangesCh
+}
+
+func (cgr *consumerGroupRegistry) claimPartition(cid *sarama.ContextID, topic string, partition int32, cancelCh <-chan none) func() {
+	if !retry(func() error { return cgr.groupMemberZNode.ClaimPartition(topic, partition) }, nil,
+		fmt.Sprintf("<%s> failed to claim partition", cid), cgr.config.Consumer.BackOffTimeout, cancelCh,
+	) {
+		log.Infof("<%s> partition claimed", cid)
+	}
+	return func() {
+		if !retry(func() error { return cgr.groupMemberZNode.ReleasePartition(topic, partition) },
+			func(err error) bool { return err != nil && err != kazoo.ErrPartitionNotClaimed },
+			fmt.Sprintf("<%s> failed to release partition", cid), cgr.config.Consumer.BackOffTimeout, cancelCh,
+		) {
+			log.Infof("<%s> partition released", cid)
+		}
+	}
+}
+
+func (cgr *consumerGroupRegistry) stop() {
+	close(cgr.stoppingCh)
+	cgr.wg.Wait()
+}
+
+// partitionOwner returns the id of the consumer group member that has claimed
+// the specified topic/partition.
+func (cgr *consumerGroupRegistry) partitionOwner(topic string, partition int32) (string, error) {
+	owner, err := cgr.groupZNode.PartitionOwner(topic, partition)
+	if err != nil {
+		return "", err
+	}
+	if owner == nil {
+		return "", nil
+	}
+	return owner.ID, nil
+}
+
+// watcher keeps an eye on the consumer group membership changes, that is when
+// members join/leave the group and sends notifications about such changes down
+// to the `membershipChangesCh` channel.
+func (cgr *consumerGroupRegistry) watcher() {
+	cid := cgr.baseCID.NewChild("watcher")
+	defer cid.LogScope()()
+	defer close(cgr.membershipChangesCh)
+
+	if cgr.retry(cgr.groupZNode.Create, nil, fmt.Sprintf("<%s> failed to create a group znode", cid)) {
+		return
+	}
+
+	var nilOrDelayCh <-chan time.Time
+watchLoop:
+	for {
+		log.Infof("<%s> retrieving group membership state...", cid)
+		var members []*kazoo.ConsumergroupInstance
+		var membershipChangedCh <-chan zk.Event
+		if cgr.retry(
+			func() error {
+				var err error
+				members, membershipChangedCh, err = cgr.groupZNode.WatchInstances()
+				return err
+			},
+			nil, fmt.Sprintf("<%s> failed to watch members", cid),
+		) {
+			return
+		}
+		// To avoid unnecessary rebalancing in case of a deregister/register
+		// sequences that happen when a member updates its topic subscriptions,
+		// we delay notification a little bit.
+		nilOrDelayCh = time.After(cgr.config.Consumer.RebalanceDelay)
+		select {
+		case <-membershipChangedCh:
+			continue watchLoop
+		case <-nilOrDelayCh:
```

Comment:
This channel is never nil when this select statement executes, because it is set to `time.After(cgr.config.Consumer.RebalanceDelay)` immediately before the select.